### DISCUSSION TITLE
Login providers & user storage

### DISCUSSION THREAD

**Discussion by zofrex** [andrew-d/homeauth#17] [General]
This is one area where the code and configuration are currently not simple, and definitely a lot less simple than I was envisioning.

My original idea for Better Than Basic was to only have username+password as a login option, and to store users and password hashes in a separate file to the configuration file, similar to how nginx and Apache keep credentials for basic auth in a .htpasswd file. Something like this:

```
user1:$apr1$/woC1jnP$KAh0SsVn5qeSMjTtn0E9Q0
user2:$apr1$QdR8fNLT$vbCEEzDj7LyqCMyNpSoBh/
user3:$apr1$Mr5A0e.U$0j39Hp5FfxRkneklXaMrr/
```

This means:

* Code is small and simple: There's only one user lookup branch, one login checking branch, etc etc
* Adding users by hand is very easy, just add a new line to the file
* It's also easy to manage users via configuration management tooling, one line per user
* UX on the login page is simple (there's one way to log in, you don't have to remember which one to use)

homeauth supports more login options than BTB, namely, username+password, email magic links, and passkeys (and maybe sign in with Google?). All of these are live at once, so different users could be signing in via different methods.

The users — as befits the complex (relative to two fields in htaccess) data structures they need to be to support all these — are stored in the "database" along with all the other data.

The downsides of this are:

* Code is larger and more complex: The login handling has to branch based on the authentication method chosen
* Adding users by hand is not super fun, as they're in the middle of a large blob of json
* Adding users via cfgmgmt tools would require parsing the json and writing it back, which is generally a lot more steps than doing line-based editing
* UX on the login page is less simple, users will need to remember which login type they need (or we'll need to make it two-step where they give the username first, and then we tell them how they auth)

The upside is flexibility, I presume. Your technologically advanced users could use passkeys while the less early adopter people could use passwords and that one person who keeps forgetting their passwords can use magic links.

Personally I'm concerned about the complexity (code and UX) and unsure if the upside is worth it. I'm actually kinda nervous about supporting _any_ configurability rather than just picking one option, from a focus and security point of view.

On the other hand, I don't think I can make a strong enough argument here to get rid of all the login options except one. They each have their merits, for one. For two if you were going to pick one and only one it would probably be passwords, and we all agree that's the worst one. I'd love to move on from passwords.

So here's my idea to mitigate _some_ of the complexity while retaining _much_ of the flexibility: Make which login method you use a configuration parameter for homeauth, and all users have to use that one.

"But doesn't that retain all the code complexity?"

No, because we could split the branching code out into separate login providers, all implementing a common interface. There would be a fair bit of duplication between the providers, but what we'd lose in duplication we'd gain in lower complexity. Each provider would have fewer code paths than we do now, so auditing the flows through the login code would be much easier.

And only one provider would ever be in play at a time — part of the complexity with flexibility is weird interactions e.g. signing up with one method, logging in with another (or like the ol' hmac/rsa switcheroo in JWTs).

Obviously the login page UX problem would be solved, show the one button for the one method that's enabled.

And user storage could in most cases now be a single file as the structure would be way simpler and, in most cases, unchanging. E.g. if you choose passwords your user configuration file would look like an htaccess file:

```
user1:$apr1$/woC1jnP$KAh0SsVn5qeSMjTtn0E9Q0
user2:$apr1$QdR8fNLT$vbCEEzDj7LyqCMyNpSoBh/
user3:$apr1$Mr5A0e.U$0j39Hp5FfxRkneklXaMrr/
```

And if you use magic links, the file would be a list of email addresses:

```
bob@example.com
alice@example.com
```

And if you use passkeys... maybe that still needs to be in the big DB instead, I don't know.

Obviously the downside is maybe you want to use passkeys but not all your users can use them.

But I think in many cases the numbers of users will be very, very small (number of users in my htpasswd file: 4), so the chances of having hugely varying requirements between them are lower than if it was, say, 1000 (and if it's 1000 then you should probably be using a real IDP!). And it's slightly weird to give operators the decision rather than users, but with small numbers of users that the operator almost certainly knows personally, they will likely be in a good position to judge what works for their users.

And the upside of keeping things small and simple is huge.

**Comment by andrew-d** [comment]
Hm, lots of really good points and things to think about here. Here's my initial thoughts, though I definitely want to think on this:

> Code is small and simple: There's only one user lookup branch, one login checking branch, etc etc

Yes, this is definitely a thing that I don't love about the current approach... though I mitigated it a bit in the current implementation by splitting all the logic early (the `via` parameter), it's still not trivial code.

> Adding users by hand is very easy, just add a new line to the file
> It's also easy to manage users via configuration management tooling, one line per user

Yeah, agreed strongly here; I don't like the current thing, and having to use `jq` to modify the "database" isn't great.

> UX on the login page is simple (there's one way to log in, you don't have to remember which one to use)

This would be 💯

> The upside is flexibility, I presume. Your technologically advanced users could use passkeys while the less early adopter people could use passwords and that one person who keeps forgetting their passwords can use magic links.

This is exactly my concern, and the reason that I wanted to support both username/password and passkeys. I will use passkeys ~everywhere, since I have like 5x Yubikeys and iCloud etc... but I have family members that don't have any of that and will want to use a username/password, and it'd be nice to not compromise the security of my account (which is also the admin account for e.g. Proxmox!) for that.

> Personally I'm concerned about the complexity (code and UX) and unsure if the upside is worth it. I'm actually kinda nervous about supporting any configurability rather than just picking one option, from a focus and security point of view.

Also agreed that we should have minimal configuration knobs, FWIW. 

> For two if you were going to pick one and only one it would probably be passwords, and we all agree that's the worst one. I'd love to move on from passwords.

Yeah, an explicit goal here is to _not_ force users to use passwords. I never want to manage new passwords again, if I can.

> So here's my idea to mitigate some of the complexity while retaining much of the flexibility: Make which login method you use a configuration parameter for homeauth, and all users have to use that one.

💡 Huh. Okay, that's something to think pretty hard about. I like the idea, conceptually, since as you say, it dramatically simplifies both the code complexity and the login UX. On the other hand, it doesn't solve the "I want to use passkeys for my account, the admin account, but my $FAMILY_MEMBER doesn't have one" problem, which was part of the reason that I implemented my current "all of the above" approach.

At the very least, I think we can say that for a single user, they can only use a single login mechanism; i.e. alice@example.com can only log in with a password, **or** a passkey, **or** be allowed to use a magic link... but not more than one at a time (unlike today).

If we did this, we could do the two-stage login approach: enter an email address, click "login", and then either (1) get a page that asks for a password, (2) get a page saying "check your email", or (3) get prompted (without a reload?) for a passkey and then logged in automatically.

> And user storage could in most cases now be a single file as the structure would be way simpler and, in most cases, unchanging. [...]

Agreed with all of this. Certainly for passwords and magic email links, splitting that out to a line in a plain text file makes 100% sense and I'd be 👍 to that; do you have any opinions on whether we configure this in the "database" (e.g. `{"Config":{"UsersFile":"/var/lib/whatever.txt"}}`) or just say "it's `users.txt` next to the `.json` file"?

For passkeys, we have to support updates since that's what detects credential cloning; every time you authenticate, you have to update the credential in the DB. It looks like this (redacted) locally:
```json
{
  "WebAuthnCreds": {
    "aaaaaaaa-aaaa-aaaa-aaaa-aaaaaaaaaaaa": [
      {
        "id": "aaaaaaaaaaaaaaaaaaaaaaaaaaa=",
        "publicKey": "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa=",
        "attestationType": "none",
        "transport": null,
        "flags": {
          "userPresent": true,
          "userVerified": true,
          "backupEligible": true,
          "backupState": true
        },
        "authenticator": {
          "AAGUID": "AAAAAAAAAAAAAAAAAAAAAA==",
          "signCount": 0,
          "cloneWarning": false,
          "attachment": "platform"
        },
        "attestation": {
          "clientDataJSON": "bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb=",
          "clientDataHash": "ccccccccccccccccccccccccccccccccccccccccccc=",
          "authenticatorData": null,
          "publicKeyAlgorithm": 0,
          "object": "ddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddd="
        },
        "UserUUID": "aaaaaaaa-aaaa-aaaa-aaaa-aaaaaaaaaaaa",
        "FriendlyName": "Foo Bar"
      }
    ]
  }
}
```

> But I think in many cases the numbers of users will be very, very small (number of users in my htpasswd file: 4), so the chances of having hugely varying requirements between them are lower than if it was, say, 1000

Agreed in the general sense, but at least in my deployment (number of users needed: about 6), I already have the case where I want to use passkeys but another user doesn't have that option, so I'd either have to (a) use username/password myself, or (b) try and teach an elderly family member how to use Passkeys to get access to something I'm sharing with them, which is ... unlikely to work well.

**Comment by andrew-d** [comment]
@zofrex Okay, I thought about this a bit more, and I think that at least for my personal use-case, I need to support multiple login providers simultaneously. However, I had an idea that, while not _quite_ as easy as "there's one option and it's global", is still better than the current morass:

1. All login providers are required to use a single dedicated file for configuration and state storage, outside of the main database.
2. You configure one or more login providers in the main configuration file.
3. We have a very straightforward bit of code that, when logging in a user, iterates through *all* configured login providers and asks them if they know about a given username. There are three cases:
    - No login providers know about the username: we just return an error
    - Exactly one login provider knows about the username: proceed with that login provider
    - More than one login provider knows about the username: hard error, user cannot log in.

This ensures that, in the base "just passwords" case, the logic for logging in a user is no more complicated than if we'd just implemented that approach and nothing else. If there are multiple login providers, they are all completely disjoint and we don't have to worry about migrating users between them, state or session confusion, etc.

And relatedly, I was already thinking of having a "login provider" abstraction _anyway_ since I wanted to be able to wrap that up in a nice interface 😅 

Thoughts? Some half-baked examples:

Interface:
```go
type LoginProvider struct {
	Name() string
	CanHandleUsername(username string) bool
	LoginUser(username string, data []byte) error
}
```

Main config file (very WIP):
```json
{
  "Config": {
    "LoginProviders": [
      {"Name": "password", "Config": {"PasswordFile": "/var/lib/homeauth/passwords.txt"}},
      {"Name": "passkey", "Config": {"StateFile": "/var/lib/homeauth/passkeys.json"}}
    ]
  }
}
```

Open questions: how does this work with the frontend? We could have a LoginProvider also expose a method for the two HTTP handlers that it requires, one to show the login page and one to handle the POST? And then the main login page has a username field and a "Login" button which does the `CanHandleUsername` lookup and redirects to the per-`LoginProvider` handlers:

```go
type LoginProvider struct {
	/// ... above

	// auto-registered at GET /login/<Name>
	HandleShowLogin(w http.ResponseWriter, r *http.Request)

	// auto-registered at POST /login/<Name>
	HandlePostLogin(w http.ResponseWriter, r *http.Request)
}
```

**Comment by zofrex** [comment]
OK, yeah, I think that lets us have our cake and eat it to a certain degree. I like it.

I think configuration could include explicitly enabling the desired methods (and paths to the user storage could have sensible defaults and be overridable, to avoid it feeling redundant). Something like:

```
## Authentication methods to enable
## Uncomment any you want to use, comment out any you don't
auth_methods:
  - password
#  - passkey
#  - email

## Paths to user list files. Uncomment to change from default locations
# password_users: /etc/homeauth/users.password
# passkey_users: /etc/homeauth/users.passkey
# email_users: /etc/homeauth/users.email
```

That way if they're not in play we don't even load them when we start up, it's super clear what's in play and what isn't.

### Front end considerations

As for the frontend: Everyone hates two-step login forms. I'd love to avoid that where possible.

Obviously if only 'password' is enabled then we just have a normal username/password form and sidestep the issue. (This is the nice thing about configuring supported methods explicitly)

Having both password and email enabled is a tricky one. I have two ideas on how to make the UX good for both contingents:

#### Switch based on email vs user

We could say that all users logging in with password use a username, _not_ an email. Then it'd be easy to distinguish which login type is happening as soon as the user field is filled out. So the logic would be like:

* Display both username+password field
* If user enters a "@" in the username field then hide the password field and login button, and show the "login with email" button

I think I like the idea of username, not email, + password anyway as it underscores that there's no email password reset etc.

#### Front-end knowledge of users

My other idea is slightly more radical and takes advantage of the tiny scale this will be operating at: we could ship a list of hashed usernames/emails to the frontend and what login is associated with each of them, so when the user field is filled out the frontend will already know, without a round trip, what login type it needs to do.

### Misc thoughts on configuration

I know we said separate files for each user type... but I am thinking _maybe_ we could simplify and have one file for all with a very simple syntax. I'm not sure if this is the kind of simplification that makes things easier, or if it'll make it conceptually harder for people to understand. But I was thinking maybe something like this:

```
username1:$apr1$/woC1jnP$KAh0SsVn5qeSMjTtn0E9Q0
username2:$apr1$QdR8fNLT$vbCEEzDj7LyqCMyNpSoBh/
username3:$apr1$Mr5A0e.U$0j39Hp5FfxRkneklXaMrr/
someemail@example.com
someotheremail@example.com
username4:passkey
```

Where username1-3 login via passwords, the users with "@" in them log in via email, and username4 with a passkey (we'd still store the actual passkey data in the "database", but we could join that data based on username).

Another possibility would be to make it fully explicit, rather than implicit:

```
username1:password:$apr1$/woC1jnP$KAh0SsVn5qeSMjTtn0E9Q0
username2:password:$apr1$QdR8fNLT$vbCEEzDj7LyqCMyNpSoBh/
username3:password:$apr1$Mr5A0e.U$0j39Hp5FfxRkneklXaMrr/
someemail@example.com:email
someotheremail@example.com:email
username4:passkey
```

### Passkeys

I really need to learn more about how passkeys work on the backend.

One observation: With your proposal, passkeys as 2FA wouldn't be possible, only as the sole login method. I'm fine with that. Just wanted to call it out explicitly.

Also with passkey users do they even _need_ a username? The passkey can identify who they are as well as authenticate, so... maybe not? Although then how to configure them is an interesting challenge. I guess maybe by ID rather than username?

Also not sure how to handle the frontend problem if passkeys and other options are in play. Maybe there's no getting around a separate "login with passkey" button? Is there any way for us to ask the browser if the user has a passkey for the site, or even if they use passkeys at all?
